### PR TITLE
[libaec] Add new port

### LINKED ISSUE #38073: [hdf5] Disable szip by default
### Is your feature request related to a problem? Please describe.

`szip` is free to use to decode data but needs a commercial license to encode in szip format.

### Proposed solution

Remove `szip` from `default-features` in `vcpkg.json`

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bansan85**
Fixes #38073 #19125

Add libaec and replace szip by libaec in hdf5

**Add libaec**

- [X] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [X] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [X] The versioning scheme in `vcpkg.json` matches what upstream says.
- [X] The license declaration in `vcpkg.json` matches what upstream says.
- [X] The installed as the "copyright" file matches what upstream says.
- [X] The source code of the component installed comes from an authoritative source.
- [X] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is in the new port's versions file.
- [X] Only one version is added to each modified port's versions file.

**Update hdf5**

- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [X] SHA512s are updated for each updated download.
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version.~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [X] Any patches that are no longer applied are deleted from the port's directory.
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by bansan85** [comment]
Need to see which and why other packages failed

**Comment by dg0yt** [comment]
vpckg is weak at handling alternatives, and libaec is an alternative to szip. Each file name can only be owned by one port. Switching some ports to libaec while szip is still used by others will not work well. 

At least `libaec` is active and keeping up it is promise of providing a drop-in replacement. IDK if the vcpkg owners are willing to accept a full switch.

**Comment by bansan85** [comment]
From what I saw, hdf5 is the only one that use directly szip. One other package uses szip but only with hdf5.
I made few PR and they all need some work. I will fix this PR within few days.

**Comment by bansan85** [comment]
Do you think I should remove szip package ?

**Comment by JonLiu1993** [comment]
> Do you think I should remove szip package ?

Don't delete a port casually. This will cause problems for the ports that depend on it.

**Comment by JonLiu1993** [comment]
@bansan85, I want to test your features and usage locally, but I tried twice and failed to download libaec-v1.1.3.tar.gz correctly. Could you try the following locally and see if it can be downloaded normally?
````
PS F:\Feature-test\libaec> ./vcpkg install libaec:x64-windows
Computing installation plan...
The following packages will be built and installed:
    libaec:x64-windows@1.1.3
Detecting compiler hash for triplet x64-windows...
Compiler found: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.39.33519/bin/Hostx64/x64/cl.exe
Restored 0 package(s) from C:\Users\v-zhli17\AppData\Local\vcpkg\archives in 184 us. Use --debug to see more details.
Installing 1/1 libaec:x64-windows@1.1.3...
Building libaec:x64-windows@1.1.3...
-- Downloading https://gitlab.dkrz.de/k202009/libaec/-/archive/v1.1.3/libaec-v1.1.3.tar.gz -> k202009-libaec-v1.1.3.tar.gz...
[DEBUG] To include the environment variables in debug output, pass --debug-env
[DEBUG] Trying to load bundleconfig from F:\Feature-test\libaec\vcpkg-bundle.json
[DEBUG] Failed to open: F:\Feature-test\libaec\vcpkg-bundle.json
[DEBUG] Bundle config: readonly=false, usegitregistry=false, embeddedsha=nullopt, deployment=Git, vsversion=nullopt
[DEBUG] Metrics enabled.
[DEBUG] Feature flag 'binarycaching' unset
[DEBUG] Feature flag 'compilertracking' unset
[DEBUG] Feature flag 'registries' unset
[DEBUG] Feature flag 'versions' unset
[DEBUG] Feature flag 'dependencygraph' unset
Downloading https://gitlab.dkrz.de/k202009/libaec/-/archive/v1.1.3/libaec-v1.1.3.tar.gz
warning: Download failed -- retrying after 1000ms
warning: Download failed -- retrying after 2000ms
warning: Download failed -- retrying after 4000ms
error: Failed to download from mirror set
error: https://gitlab.dkrz.de/k202009/libaec/-/archive/v1.1.3/libaec-v1.1.3.tar.gz: WinHttpSendRequest failed with exit code 12002
error: https://gitlab.dkrz.de/k202009/libaec/-/archive/v1.1.3/libaec-v1.1.3.tar.gz: WinHttpSendRequest failed with exit code 12002
error: https://gitlab.dkrz.de/k202009/libaec/-/archive/v1.1.3/libaec-v1.1.3.tar.gz: WinHttpSendRequest failed with exit code 12002
error: https://gitlab.dkrz.de/k202009/libaec/-/archive/v1.1.3/libaec-v1.1.3.tar.gz: WinHttpSendRequest failed with exit code 12002
[DEBUG] D:\a\_work\1\s\src\vcpkg\base\downloads.cpp(1031):
[DEBUG] Time in subprocesses: 0us
[DEBUG] Time in parsing JSON: 2us
[DEBUG] Time in JSON reader: 0us
[DEBUG] Time in filesystem: 1131us
[DEBUG] Time in loading ports: 0us
[DEBUG] Exiting after 1.5 min (91181618us)

CMake Error at scripts/cmake/vcpkg_download_distfile.cmake:32 (message):

      Failed to download file with error: 1
      If you are using a proxy, please check your proxy setting. Possible causes are:

      1. You are actually using an HTTP proxy, but setting HTTPS_PROXY variable
         to `https://address:port`. This is not correct, because `https://` prefix
         claims the proxy is an HTTPS proxy, while your proxy (v2ray, shadowsocksr
         , etc..) is an HTTP proxy. Try setting `http://address:port` to both
         HTTP_PROXY and HTTPS_PROXY instead.

      2. If you are using Windows, vcpkg will automatically use your Windows IE Proxy Settings
         set by your proxy software. See https://github.com/microsoft/vcpkg-tool/pull/77
         The value set by your proxy might be wrong, or have same `https://` prefix issue.

      3. Your proxy's remote server is out of service.

      If you've tried directly download the link, and believe this is not a temporary
      download server failure, please submit an issue at https://github.com/Microsoft/vcpkg/issues
      to report this upstream download server failure.


Call Stack (most recent call first):
  scripts/cmake/vcpkg_download_distfile.cmake:270 (z_vcpkg_download_distfile_show_proxy_and_fail)
  scripts/cmake/vcpkg_from_gitlab.cmake:113 (vcpkg_download_distfile)
  ports/libaec/portfile.cmake:1 (vcpkg_from_gitlab)
  scripts/ports.cmake:175 (include)
````

**Comment by bansan85** [comment]
@JonLiu1993 The official website is down... https://gitlab.dkrz.de

**Comment by bansan85** [comment]
@JonLiu1993 Website is back. But "We currently have serious issues with a central network switch causing failure of most of our virtual environment. This affects most of our services."

**Comment by dg0yt** [comment]
> Don't delete a port casually. This will cause problems for the ports that depend on it.

This answer may be a little short. In the light of potential conflicts.

**Comment by JonLiu1993** [comment]
Tested libaec and hdf5 usage successfuly by `libaec:x64-windows` and `hdf5:x64-windows`

**Review by dg0yt — CHANGES_REQUESTED** [review]
> Tested libaec and hdf5 usage successfuly by `libaec:x64-windows` and `hdf5:x64-windows`

Sure. 
But do you understand what happens in full CI with these files being installed:
~~~
libaec:x64-windows:/bin/szip.dll
libaec:x64-windows:/include/szlib.h
libaec:x64-windows:/lib/szip.lib
~~~
Hint:
~~~
szip:x64-windows:/bin/szip.dll
szip:x64-windows:/include/szlib.h
szip:x64-windows:/lib/szip.lib
~~~

**Comment by JonLiu1993** [comment]
@bansan85, It seems that the above header files are installed in the wrong location. You need to take a look.

**Comment by dg0yt** [comment]
> @bansan85, It seems that the above header files are installed in the wrong location. You need to take a look.

No.
The point that already was made is that `libaec` is intented to be used as a drop-in *replacement* for szip.
I cannot assess how much this is true. The full file list indicates that it might need additional adjustments.
But it implies that vcpkg cannot build both in CI. Either we start to "skip" szip (as done with opencv2 or opencv2), or we make it an empty port which depends on `libaec`.

File: ports/libaec/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                    "${CURRENT_PACKAGES_DIR}/lib/szip-static.lib"
+                    "${CURRENT_PACKAGES_DIR}/debug/lib/aec-static.lib"
+                    "${CURRENT_PACKAGES_DIR}/debug/lib/szip-static.lib")
+    else()
+        file(REMOVE "${CURRENT_PACKAGES_DIR}/lib/libsz${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}"
+                    "${CURRENT_PACKAGES_DIR}/lib/libaec${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}"
+                    "${CURRENT_PACKAGES_DIR}/debug/lib/libsz${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}"
+                    "${CURRENT_PACKAGES_DIR}/debug/lib/libaec${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}")
+    endif()
+endif()
```
Comment (dg0yt) [inline ports/libaec/portfile.cmake:50]: I really think this should be controlled by a patch. Otherwise CMake config might refer to missing files.

**Comment by bansan85** [comment]
I think the location of files are good. But it's true that there is a conflict with szip.

I replaced the deletion of unused library and patched CMakeLists.txt

**Comment by bansan85** [comment]
I just noticed caffe2:

portfile.cmake:

```
SET(VCPKG_POLICY_EMPTY_PACKAGE enabled)
message(WARNING "The 'caffe2' port is deprecated, using 'libtorch' instead")
```

vcpkg.json:

```
{
  "name": "caffe2",
  "version": "0.8.1",
  "port-version": 8,
  "description": "Caffe2 is a lightweight, modular, and scalable deep learning framework.",
  "homepage": "https://github.com/caffe2/caffe2",
  "license": "Apache-2.0",
  "dependencies": [
    {
      "name": "libtorch",
      "default-features": false
    }
  ]
}
```

Maybe I could increase port-version of szip and deprecate it ?

**Comment by dg0yt** [comment]
This is an example of "make it an empty port which depends on" the new port. This requires a (mostly) complete drop-in replacement.

The downside of this transition is that no new version of szip could be added to the vcpkg registry as long as it overlaps with libaec. OTOH there is no demand for szip in the vcpkg registry except for hdf5, and they prefer libaec IIRC. The benefit of choosing the preferred implementation is important enough IMO.

The alternative is to patch libaec and hdf5 to use non-conflicting names. And this would still come with extra efforts to avoid symbol clashes with msbuild autolinking. I hope this will not be asked for.

(You really take an adavanced hands-on course on vcpkg porting with this PR. :-)

**Review by JonLiu1993 — DISMISSED** [review]

**Comment by JonLiu1993** [comment]
Tested libaec and hdf5 usage successfuly by libaec:x64-windows and hdf5:x64-windows

**Comment by bansan85** [comment]
New week, better motivation :)

It's good to know that the patent has been removed.

I know you already answered but please confirm me which solution you want:

  - Remove szip and use libaec,
  - Fake szip and use libaec. Add a warning message if szip is used. If someone complains, implement one of the solutions below.
  - Add support for both as feature (only one should be enabled at the same time). Both libraries can't be installed at the same time (no patch to rename library filename),
  - Add support for both as feature (only one should be enabled at the same time). Both libraries can be installed at the same time (need to patch to rename library filename). Should I rename szip library or libaec library ?

**Comment by vicroms** [comment]
@bansan85 

The team discussed this PR and we think that changing all uses of `szip` to `libaec` seems like the best path. Could you please make the necessary changes to replace all uses of `szip` and remove the port.

File: versions/s-/szip.json [hunk 2]
```diff
@@ -1,79 +0,0 @@
-{
```
Comment (JonLiu1993) [inline versions/s-/szip.json:1]: Don't delete it, it should be left to those who own the source asset and pin the version.

**Comment by JonLiu1993** [comment]
Tested libaec and hdf5 usage successfuly by libaec:x64-windows and hdf5:x64-windows

**Review by dg0yt — CHANGES_REQUESTED** [review]
This is not ready for merge.

Must-have: CMake config must support multi-config. (It is not an exported config!)

Nice-to-have: Trim the patch by not changing indentation.

**Comment by bansan85** [comment]
> This is not ready for merge.
> 
> Must-have: CMake config must support multi-config. (It is not an exported config!)

I don't get it.

For libaec, I have:

```
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
```

Is this for hdf5 that have 2 `Configuring` that should be merged into one ? 

```
-- Configuring x64-windows-dbg
-- Configuring x64-windows-rel
-- Building x64-windows-dbg
-- Building x64-windows-rel
```

> 
> Nice-to-have: Trim the patch by not changing indentation.

So you would like I add if()/endif() and not indent inside it ? OK.

**Comment by dg0yt** [comment]
> > Must-have: CMake config must support multi-config. (It is not an exported config!)

It is not about how libaec is built.
It is about the CMake config package installed by the port:
~~~
libaec:x64-windows:/share/libaec/libaec-config-version.cmake
libaec:x64-windows:/share/libaec/libaec-config.cmake
~~~
There are not the usual files with exported target properties for release and debug. AFAICS `libaec-config.cmake` is hand-written, as old find modules did. By using this config will not get multi-config variables and target properties. They may end up with mixing debug and release binaries, which means incompatible runtimes on Windows.

So either `libaec-config.cmake` needs to be patched (or replaced) for multi-config support (`select_library_configurations`, `IMPORTED_LOCATION_<CONFIG>`), or it is replaced with a proper export of the CMake targets. (The second option is the prefered solution also for upstreaming. But if you are not familiar with that, it might be a quite some effort.)

For reference, `szip` had a typical exported config package:
~~~
szip:x64-windows:/share/szip/szip-config-version.cmake
szip:x64-windows:/share/szip/szip-config.cmake
szip:x64-windows:/share/szip/szip-targets-debug.cmake
szip:x64-windows:/share/szip/szip-targets-release.cmake
szip:x64-windows:/share/szip/szip-targets.cmake
~~~

**Comment by bansan85** [comment]
@dg0yt I implemented the export of targets using CMake features and cleanup `libaec-config.cmake`.

If this is fine I will report in a second time the patch to upstream.

**Comment by JonLiu1993** [comment]
@bansan85, could you fix the conflicting file?

**Comment by JonLiu1993** [comment]
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

**Comment by nickanthony-dgl** [comment]
I'm getting build errors in HDF5 1.14.4.3 port 3. I would like to try rolling back to older versions of the port but it seems that I can't do that because: "error: the baseline does not contain an entry for port szip". 

Why was this merged?

**Comment by bansan85** [comment]
> Why was this merged?

Because it worked. But it's no possible to check all existing configuration. Can you create an issue with your log and ping me ? Thanks.

**Comment by nickanthony-dgl** [comment]
The build error I have is related to HDF5 14.4.3, I will open a separate issue for that. The point is that this PR made it so that there is no way to use any version of the HDF5 port except for the one introduced in this PR unless you use a baseline that is before this PR was introduced.

Adding something like this to your manifest:
```
"overrides": [
  {
    "name": "hdf5",
    "version": "1.14.2",
  }
]
```
will result in "error: the baseline does not contain an entry for port szip"

**Comment by bansan85** [comment]
OK. I got it. Thank. I will revert my commit https://github.com/microsoft/vcpkg/pull/38107/commits/4d4da8d47dbf352efba690418bde7699c85cd531 .

**Comment by dg0yt** [comment]
@nickanthony-dgl Did you try to add an `overrides` for `szip`, too?

**Comment by nickanthony-dgl** [comment]
I didn't think to try that. I had assumed that if there was no entry for szip in the baseline then vcpkg wouldn't understand what was being overridden. For now I am just sticking with a baseline from before this PR but I can try that later.

**Comment by bansan85** [comment]
> @nickanthony-dgl Did you try to add an `overrides` for `szip`, too?

Yes, adding an overrides for szip fix the problem:

```
{
    "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg-tool/main/docs/vcpkg.schema.json",
    "builtin-baseline": "f523208df4ae467436dc2d2eeab594fed240d552",
    "dependencies": [
        "hdf5"
    ],
    "overrides": [
        {
            "name": "hdf5",
            "version": "1.14.2",
            "port-version": 0
        },
        {
            "name": "szip",
            "version": "2.1.1",
            "port-version": 12
        }
    ]
}
```

But I understand this is really annoying.

@dg0yt Are you fine if I revert https://github.com/microsoft/vcpkg/commit/4d4da8d47dbf352efba690418bde7699c85cd531 ?

**Pull request merged**